### DISCUSSION TITLE
No cobertura files for ms code coverage.

### DISCUSSION THREAD

**Discussion by alejandroemoreno** [FortuneN/FineCodeCoverage#296] [Q&A] [answered]
Can't get any coverage results.

I've pared the test project down to the minimum.

* Used to have coverlet.collector package - removed it
* Used to have coverlet.runsettings file - removed it
* Added Microsoft.CodeCoverage

Still nothing.

This used to work several months ago. When it stopped working I did not have time to dig into it. 

* Tests are running fine
* FCC 1.1.180
* Microsoft.CodeCoverage 17.3.1
* Microsoft.NET.Test.Sdk 17.3.1

FCC output:
```
Fine Code Coverage : Ms code coverage

Fine Code Coverage : No cobertura files for ms code coverage.

Fine Code Coverage : ================================== DONE ==================================
```

Help?

**Comment by tonyhallett** [comment]
What is your RunMsCodeCoverage setting ?

**Comment by alejandroemoreno** [comment]
It's set to *Yes*
![image](https://user-images.githubusercontent.com/89926271/190499748-061615d3-5de5-4ae5-aa06-e2b78ae36220.png)

**Comment by tonyhallett** [comment]
Are you supplying a runsettings ?

**Reply by tonyhallett** [reply]
Can you paste the output from the FCC output window pane.

**Reply by alejandroemoreno** [reply]
No runsettings. I tried to create one but that only resulted in various other errors, so I decided to remove it altogether.

The FCC output is just as I pasted it originally:
```

Fine Code Coverage : Ms code coverage with user runsettings

Fine Code Coverage : No cobertura files for ms code coverage.

Fine Code Coverage : ================================== DONE ==================================

Fine Code Coverage : Ms code coverage with user runsettings

Fine Code Coverage : No cobertura files for ms code coverage.

Fine Code Coverage : ================================== DONE ==================================

Fine Code Coverage : Ms code coverage

Fine Code Coverage : No cobertura files for ms code coverage.

Fine Code Coverage : ================================== DONE ==================================

Fine Code Coverage : Ms code coverage

Fine Code Coverage : No cobertura files for ms code coverage.

Fine Code Coverage : ================================== DONE ==================================

Fine Code Coverage : Ms code coverage

Fine Code Coverage : No cobertura files for ms code coverage.

Fine Code Coverage : ================================== DONE ==================================
```

You can see the first two runs when I still had a runsettings.

**Reply by tonyhallett** [reply]
Of course, apologies.

The error is specific and as yet has not been encountered. The test explorer should provide the location of the cobertura file but has failed to do so.

Is there anything unusual with your solution? Perhaps you can share it or a stripped down version of it.

**Reply by alejandroemoreno** [reply]
Nothing that I can think of.

The file *is* being created:
![image](https://user-images.githubusercontent.com/89926271/190512369-4a87d6be-0d4b-47c2-8353-342dd2f3176f.png)

How does Test Explorer expose the cobertura file?
How does FCC look for it?

**Reply by alejandroemoreno** [reply]
(I'd have to put in some work to pare down the solution into a shareable state ...)

**Reply by tonyhallett** [reply]
Exposed with API.

https://github.com/FortuneN/FineCodeCoverage/blob/ca33352a053228bba4d4285a8c2efdb0ea529bda/SharedProject/Core/MsTestPlatform/CodeCoverage/MsCodeCoverageRunSettingsService.cs#L352

**Reply by alejandroemoreno** [reply]
So maybe test explorer is reporting the location wrong? Any way to control that?

**Comment by tonyhallett** [comment]
It is reporting no location which is odd given you can find it. 
Can you delete the Test results folder and try again. What does FCC report and is a cobertura file created ?

**Reply by tonyhallett** [reply]
I am also interested in the class.js file. Any ideas what that is ?

**Reply by alejandroemoreno** [reply]
The beginning of class.js:
```js
/* Chartist.js 0.11.4
 * Copyright © 2019 Gion Kunz
 * Free to use under either the WTFPL license or the MIT license.
 * https://raw.githubusercontent.com/gionkunz/chartist-js/master/LICENSE-WTFPL
 * https://raw.githubusercontent.com/gionkunz/chartist-js/master/LICENSE-MIT
 */
```

**Reply by tonyhallett** [reply]
Peculiar..

**Reply by alejandroemoreno** [reply]
Removing the test results and running the tests again did NOT generate coverage results. Those files must have been from a previous attempts earlier today.

FCC again responded with "No cobertura files ...".

I think there's a chance my test project is configured funny. Seriously considering creating a new test project and copy my .cs files over.

**Reply by tonyhallett** [reply]
Can you restart you machine and try again ?

**Comment by tonyhallett** [comment]
What project type is your test project ?

**Reply by alejandroemoreno** [reply]
Net6.0 console app.

**Comment by tonyhallett** [comment]
Can you provide the generated runsettings in the test project fcc folder.

**Reply by alejandroemoreno** [reply]
Can't find any runsettings file in folder `fine-code-coverage-net6.0` 🤔

**Reply by tonyhallett** [reply]
For a non multi targeted net core test project the containing folder is (normally) bin\Debug\net6.0\fine-code-coverage\coverage-tool-output.  I cannot remember how FCC determines paths but this is the norm.

I am done for the day.  I will look at your response tomorrow.

**Reply by alejandroemoreno** [reply]
Oh! Found it.

```xml
<?xml version="1.0" encoding="utf-8"?>
<RunSettings>
  <RunConfiguration>
    <ResultsDirectory>C:\Users\Alejandro\source\surrogate-filing\filing-api\tests\jsg-filing-surrogate-api.tests\bin\Debug\net6.0\fine-code-coverage\coverage-tool-output</ResultsDirectory>
    <TestAdaptersPaths>C:\Users\Alejandro\AppData\Local\FineCodeCoverage\msCodeCoverage\17.1.0\build\netstandard1.0</TestAdaptersPaths>
    <CollectSourceInformation>False</CollectSourceInformation>
  </RunConfiguration>
  <DataCollectionRunSettings>
    <DataCollectors>
      <DataCollector friendlyName="Code Coverage" enabled="True">
        <Configuration>
          <CodeCoverage>
            <ModulePaths>
              <Exclude>
                <ModulePath>C:\\Users\\Alejandro\\source\\surrogate-filing\\filing-api\\tests\\jsg-filing-surrogate-api.tests\\bin\\Debug\\net6.0\\jsg-filing-surrogate-api.tests.dll</ModulePath>
              </Exclude>
              <Include></Include>
            </ModulePaths>
            <Functions>
              <Exclude></Exclude>
              <Include></Include>
            </Functions>
            <Attributes>
              <Exclude></Exclude>
              <Include></Include>
            </Attributes>
            <Sources>
              <Exclude></Exclude>
              <Include></Include>
            </Sources>
            <CompanyNames>
              <Exclude></Exclude>
              <Include></Include>
            </CompanyNames>
            <PublicKeyTokens>
              <Exclude></Exclude>
              <Include></Include>
            </PublicKeyTokens>
            <UseVerifiableInstrumentation>True</UseVerifiableInstrumentation>
            <AllowLowIntegrityProcesses>True</AllowLowIntegrityProcesses>
            <CollectFromChildProcesses>True</CollectFromChildProcesses>
            <CollectAspDotNet>False</CollectAspDotNet>
          </CodeCoverage>
          <Format>Cobertura</Format>
          <FCCGenerated />
        </Configuration>
      </DataCollector>
    </DataCollectors>
  </DataCollectionRunSettings>
</RunSettings>
```
The paths indicated here look correct to me.

**Reply by alejandroemoreno** [reply]
How does Test Explorer know that it has to generate coverage output?

**Reply by tonyhallett** [reply]
The DataCollector element

**Reply by tonyhallett** [reply]
and the ResultsDirectory should contain the cobetura that is reported by the API

**Reply by alejandroemoreno** [reply]
Here's my .csproj for reference:

```xml
<Project Sdk="Microsoft.NET.Sdk">

    <PropertyGroup>
        <TargetFramework>net6.0</TargetFramework>
        <RootNamespace>JSGEstate.Tests</RootNamespace>

        <IsPackable>false</IsPackable>

        <AssemblyName>jsg-filing-surrogate-api.tests</AssemblyName>
    </PropertyGroup>
    <PropertyGroup>
        <UseDataCollector />
    </PropertyGroup>
    <PropertyGroup Label="FineCodeCoverage">
        <Enabled>True</Enabled>
        <Exclude></Exclude>
        <Include>*</Include>
        <ExcludeByFile>
            **/Migrations/**
            **/Resources/**
            **/DataContext/**
            **/IoC/**
            **/Entities/**
            **/Controllers/**
            **/Models/**
        </ExcludeByFile>
        <ExcludeByAttribute>MyCustomExcludeFromCodeCoverage</ExcludeByAttribute>
        <IncludeTestAssembly>False</IncludeTestAssembly>
    </PropertyGroup>
    <ItemGroup>
        <PackageReference Include="ILogger.Moq" Version="1.1.10" />
        <PackageReference Include="Microsoft.CodeCoverage" Version="17.3.1" />
        <PackageReference Include="Microsoft.EntityFrameworkCore.InMemory" Version="6.0.9" />
        <PackageReference Include="Moq.AutoMock" Version="3.4.0" />
        <PackageReference Include="nunit" Version="3.13.3" />
        <PackageReference Include="NUnit3TestAdapter" Version="4.2.1" />
        <PackageReference Include="Microsoft.NET.Test.Sdk" Version="17.3.1" />
        <PackageReference Include="System.ValueTuple" Version="4.5.0" />
    </ItemGroup>
    <ItemGroup>
        <ProjectReference Include="..\..\jsg-filing-surrogate-api\jsg-filing-surrogate-api.csproj" />
    </ItemGroup>
</Project>
```
And the ResultsDirectory does contain a folder with a GUID for a name, but it's empty.

**Reply by tonyhallett** [reply]
the API should report a cobertura file in the ResultsDirectory

**Reply by alejandroemoreno** [reply]
The API is probably doing its job, but the ResultsDirectory/{GUID} folder is empty.
I can't see *any* coverage files in the test project outputs.

**Reply by tonyhallett** [reply]
What version of visual studio do you have ? An update may be necessary.

**Reply by alejandroemoreno** [reply]
Microsoft Visual Studio Community 2022 (64-bit) - Current
Version 17.3.4

**Reply by tonyhallett** [reply]
I assume that you are not using this.  If so  please remove it and try again.
```
  <PackageReference Include="Microsoft.CodeCoverage" Version="17.3.1" />
```

**Reply by alejandroemoreno** [reply]
Remove? This?

**Reply by tonyhallett** [reply]
it was hiding - resolved

**Reply by alejandroemoreno** [reply]
Removed. No difference.
Test Explorer ran all tests, created a new {GUID} folder in the ResultsDirectory, but it has zero files.
FCC's output is pretty much the same:
```
Fine Code Coverage : Initializing

Fine Code Coverage : Initialized

Fine Code Coverage : Ms code coverage

Fine Code Coverage : No cobertura files for ms code coverage.

Fine Code Coverage : ================================== DONE ==================================
```

**Comment by alejandroemoreno** [comment, marked as answer]
Well, it looks like one of the tests is breaking the coverage.
Another dev on my team removed all tests except one, and the coverage worked.
We're slowly bringing them back into the project.
Once I figure out the culprit I'll share it here.
@tonyhallett, thank you so much for your help!

**Reply by tonyhallett** [reply]
You could use the dotnet cli with the generated runsettings and see if that logs error information that Fcc may have missed

**Reply by alejandroemoreno** [reply]
Problem solved!

Two tests made use of temp files and, during clean up, we accidentally attempted to clear the entire Temp folder. 🤦‍♂️

Once we ensured we only attempted to delete the relevant file, FCC was able to find the cobertura file. (I imagine we were accidentally clobbering some of the test runner's working files...)
